File: frontend/src/api/projects.ts
import { get, post, put, del } from './index'
import type { SprintOption } from './types'

// ⚠ 报告类接口的**实现与类型统一收敛在 @/api/reports**,本模块不再重复定义
//   (历史上这里另有一份 BugDetailResponse / BugListItem / ReopenBugItem 与
//    getBugList,签名还与 reports.ts 的参数顺序相反,属于随时会踩的坑)。
//   这里只保留 re-export,避免打断既有 `from '@/api/projects'` 的引用。
export type { ProjectOption, SprintOption, SprintMetrics } from './types'

export interface ReminderSettings {
  need_story_remind: boolean
  need_task_remind: boolean
  need_sonar_scan_remind: boolean
  need_report_data: boolean
  story_remind_time?: string
  task_remind_time?: string
  sonar_remind_time?: string
}

export interface ProjectConfig {
  id: number
  board_id: string
  board_name: string
  project_id: string
  project_name: string
  gitlab_group_key: string
  sonar_key_prefix: string
  sonar_scan_remind_default_person: string
  robot_key: string
  jira_user: string
  jira_token?: string
  created_at?: string
  updated_at?: string
  need_story_remind?: boolean
  need_task_remind?: boolean
  need_sonar_scan_remind?: boolean
  need_report_data?: boolean
  story_remind_time?: string
  task_remind_time?: string
  sonar_remind_time?: string
  reminder_settings?: ReminderSettings
}

export interface ProjectFormData {
  id?: number
  board_id: string
  board_name: string
  project_id: string
  project_name: string
  gitlab_group_key: string
  sonar_key_prefix: string
  sonar_scan_remind_default_person: string
  robot_key: string
  jira_user: string
  jira_token: string
  need_story_remind: boolean
  need_task_remind: boolean
  need_sonar_scan_remind: boolean
  need_report_data: boolean
  story_remind_time: string
  task_remind_time: string
  sonar_remind_time: string
}

/** 创建/更新项目的请求体:开关字段嵌套在 reminder_settings 中 */
export interface ProjectPayload extends Omit<ProjectFormData, 'need_story_remind' | 'need_task_remind' | 'need_sonar_scan_remind' | 'need_report_data' | 'story_remind_time' | 'task_remind_time' | 'sonar_remind_time'> {
  reminder_settings: ReminderSettings
}

export const projectApi = {
  getList(): Promise<ProjectConfig[]> {
    return get<ProjectConfig[]>('/projects')
  },

  /**
   * 单个项目详情。
   *
   * ⚠ 编辑表单必须走这里而不是复用列表行:列表接口对 robot_key 做了脱敏,
   *   拿列表值填表单再保存会把掩码写回库,真实 webhook key 被覆盖。
   *   详情接口返回明文(含 jira_token),供表单回显。
   */
  getById(id: number): Promise<ProjectConfig> {
    return get<ProjectConfig>(`/projects/${id}`)
  },

  create(data: ProjectPayload): Promise<ProjectConfig> {
    return post<ProjectConfig>('/projects', data)
  },

  update(id: number, data: ProjectPayload): Promise<ProjectConfig> {
    return put<ProjectConfig>(`/projects/${id}`, data)
  },

  delete(id: number): Promise<void> {
    return del<void>(`/projects/${id}`)
  },
}

/**
 * RDM **实时**迭代列表(作业页「手动执行」弹窗的 Sprint 下拉数据源)。
 *
 * ⚠ 与 `@/api/reports` 的 `reportsApi.getSprints` **同名但不同源**,别混用:
 *   - 本模块(这里):`GET /reports/sprints/{id}` —— 直连 RDM 实时拉取,
 *     拉取成功还会全量重写本地 `rdm_sprint`(见后端 reports.py 该端点的副作用说明);
 *     好处是新迭代**首次同步前**也能选到。
 *   - `@/api/reports`: `GET /reports/db-sprints/{id}` —— 只读本地库,不碰 RDM。
 *
 * ⚠ 导出名刻意**不叫** `reportsApi`:同名导出会让编辑器的自动导入选错文件,
 *   而两者一旦选错就是「静默换了数据源」而不是编译报错。
 */
export const sprintApi = {
  getSprints(projectId: number): Promise<SprintOption[]> {
    return get<SprintOption[]>(`/reports/sprints/${projectId}`)
  },
}

File: backend/src/api/routes/projects.py
import logging
import re

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, field_validator

from api.auth import get_current_user_from_header
from db.database import get_session
from db.models import ProjectConfig, ProjectReminderSettings

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/projects", tags=["项目配置"])


_HHMM_RE = re.compile(r"^([01]\d|2[0-3]):[0-5]\d$")
# 列表接口对 robot_key 的掩码形态(见 db/models.py 的 mask_robot_key)
_MASKED_SECRET_RE = re.compile(r"^.{4}\*{4}.{4}$")


def _looks_masked(value) -> bool:
    """判断值是否形如列表接口吐出的掩码(如 abcd****wxyz)。"""
    return isinstance(value, str) and bool(_MASKED_SECRET_RE.match(value))


def _validate_hhmm(value: str | None) -> str | None:
    """校验 HH:MM 格式，允许 None / 空字符串（表示未配置）"""
    if value is None or value == "":
        return value
    if not _HHMM_RE.match(value):
        raise ValueError("时间格式必须为 HH:MM（24小时制），如 09:30 / 18:00")
    return value


class ProjectReminderSettingsCreate(BaseModel):
    need_story_remind: bool = False
    need_task_remind: bool = False
    need_sonar_scan_remind: bool = False
    need_report_data: bool = False
    # 各任务类型的自定义调度时间 (HH:MM格式)
    story_remind_time: str | None = None
    task_remind_time: str | None = None
    sonar_remind_time: str | None = None

    @field_validator("story_remind_time", "task_remind_time", "sonar_remind_time")
    @classmethod
    def _check_time_format(cls, v: str | None) -> str | None:
        return _validate_hhmm(v)


class ProjectConfigCreate(BaseModel):
    board_id: str
    board_name: str
    project_id: str
    project_name: str
    gitlab_group_key: str = ""
    sonar_key_prefix: str = ""
    sonar_scan_remind_default_person: str = ""
    robot_key: str = ""
    jira_user: str = ""
    jira_token: str = ""
    reminder_settings: ProjectReminderSettingsCreate | None = None


class ProjectConfigUpdate(BaseModel):
    board_name: str | None = None
    board_id: str | None = None
    project_id: str | None = None
    project_name: str | None = None
    gitlab_group_key: str | None = None
    sonar_key_prefix: str | None = None
    sonar_scan_remind_default_person: str | None = None
    robot_key: str | None = None
    jira_user: str | None = None
    jira_token: str | None = None
    reminder_settings: ProjectReminderSettingsCreate | None = None


class ProjectReminderSettingsResponse(BaseModel):
    id: int
    project_config_id: int
    need_story_remind: bool
    need_task_remind: bool
    need_sonar_scan_remind: bool
    need_report_data: bool
    story_remind_time: str | None = None
    task_remind_time: str | None = None
    sonar_remind_time: str | None = None
    created_at: str | None = None
    updated_at: str | None = None


class ProjectConfigResponse(BaseModel):
    id: int
    board_id: str
    board_name: str
    project_id: str
    project_name: str
    gitlab_group_key: str
    sonar_key_prefix: str
    sonar_scan_remind_default_person: str
    robot_key: str
    jira_user: str
    need_story_remind: bool = False
    need_task_remind: bool = False
    need_sonar_scan_remind: bool = False
    need_report_data: bool = False
    story_remind_time: str | None = None
    task_remind_time: str | None = None
    sonar_remind_time: str | None = None
    created_at: str | None = None
    updated_at: str | None = None


@router.get("", response_model=list[ProjectConfigResponse])
async def list_projects(current_user: dict = Depends(get_current_user_from_header)):
    """获取项目列表(所有用户共享,返回全部项目)"""
    with get_session() as session:
        projects = session.query(ProjectConfig).all()
        return [p.to_dict(include_token=False) for p in projects]


@router.get("/{project_id}", response_model=ProjectConfigResponse)
async def get_project(
    project_id: int, current_user: dict = Depends(get_current_user_from_header)
):
    """获取单个项目详情"""
    with get_session() as session:
        project = (
            session.query(ProjectConfig).filter(ProjectConfig.id == project_id).first()
        )
        if not project:
            raise HTTPException(status_code=404, detail="项目不存在")

        # 详情接口返回 robot_key 明文:编辑表单需要真实值回显(见 to_dict 的说明)
        return project.to_dict(include_token=True, mask_robot_key=False)


@router.post("", response_model=ProjectConfigResponse)
async def create_project(
    config: ProjectConfigCreate,
    current_user: dict = Depends(get_current_user_from_header),
):
    """创建新项目配置"""
    with get_session() as session:
        # 检查 board_id 是否已存在
        existing = (
            session.query(ProjectConfig)
            .filter(ProjectConfig.board_id == config.board_id)
            .first()
        )
        if existing:
            raise HTTPException(status_code=400, detail="Board ID 已存在")

        # 检查 project_id 是否已存在
        existing = (
            session.query(ProjectConfig)
            .filter(ProjectConfig.project_id == config.project_id)
            .first()
        )
        if existing:
            raise HTTPException(status_code=400, detail="Project ID 已存在")

        # 创建项目配置
        new_project = ProjectConfig(
            board_id=config.board_id,
            board_name=config.board_name,
            project_id=config.project_id,
            project_name=config.project_name,
            gitlab_group_key=config.gitlab_group_key,
            sonar_key_prefix=config.sonar_key_prefix,
            sonar_scan_remind_default_person=config.sonar_scan_remind_default_person,
            robot_key=config.robot_key,
            jira_user=config.jira_user,
            jira_token=config.jira_token,
        )
        session.add(new_project)
        session.flush()  # 获取项目ID

        # 创建提醒设置
        if config.reminder_settings:
            rs = config.reminder_settings
            reminder_settings = ProjectReminderSettings(
                project_config_id=new_project.id,
                need_story_remind=rs.need_story_remind,
                need_task_remind=rs.need_task_remind,
                need_sonar_scan_remind=rs.need_sonar_scan_remind,
                need_report_data=rs.need_report_data,
                story_remind_time=getattr(rs, "story_remind_time", None),
                task_remind_time=getattr(rs, "task_remind_time", None),
                sonar_remind_time=getattr(rs, "sonar_remind_time", None),
            )
            session.add(reminder_settings)

        session.commit()
        session.refresh(new_project)

        return new_project.to_dict()


@router.put("/{project_id}", response_model=ProjectConfigResponse)
async def update_project(
    project_id: int,
    config: ProjectConfigUpdate,
    current_user: dict = Depends(get_current_user_from_header),
):
    """更新项目配置"""
    with get_session() as session:
        project = (
            session.query(ProjectConfig).filter(ProjectConfig.id == project_id).first()
        )
        if not project:
            raise HTTPException(status_code=404, detail="项目不存在")

        # 更新项目字段
        update_data = config.model_dump(exclude_unset=True)

        # 分离提醒设置和项目配置
        reminder_settings_data = update_data.pop("reminder_settings", None)

        # 将 reminder_settings_data 转换为 dict（如果是 Pydantic 模型）
        if reminder_settings_data is not None and hasattr(
            reminder_settings_data, "model_dump"
        ):
            reminder_settings_data = reminder_settings_data.model_dump(
                exclude_unset=True
            )

        for key, value in update_data.items():
            if key == "reminder_settings":
                continue
            # 防御:列表接口对 robot_key 脱敏,若调用方把掩码原样回传,绝不能写进库
            # (否则真实 webhook key 会被 "abcd****wxyz" 覆盖,企微推送直接失效)
            if key == "robot_key" and _looks_masked(value):
                logger.warning(
                    "忽略疑似脱敏值的 robot_key 更新: project_id=%s", project_id
                )
                continue
            if hasattr(project, key):
                setattr(project, key, value)

        # 更新提醒设置
        if reminder_settings_data is not None:
            existing_settings = (
                session.query(ProjectReminderSettings)
                .filter(ProjectReminderSettings.project_config_id == project_id)
                .first()
            )

            if existing_settings:
                for key, value in reminder_settings_data.items():
                    if hasattr(existing_settings, key):
                        setattr(existing_settings, key, value)
            elif reminder_settings_data:
                # 如果之前没有设置，创建新的
                new_settings = ProjectReminderSettings(
                    project_config_id=project_id,
                    **reminder_settings_data,
                )
                session.add(new_settings)

        session.commit()
        session.refresh(project)

        return project.to_dict()


@router.delete("/{project_id}")
async def delete_project(
    project_id: int, current_user: dict = Depends(get_current_user_from_header)
):
    """删除项目配置"""
    with get_session() as session:
        project = (
            session.query(ProjectConfig).filter(ProjectConfig.id == project_id).first()
        )
        if not project:
            raise HTTPException(status_code=404, detail="项目不存在")

        session.delete(project)
        session.commit()

        return {"message": "项目已删除"}
